signer/signer.go

```diff
@@ -0,0 +1,243 @@
+package signer
+
+import (
+	"crypto/x509"
+	"fmt"
+	"time"
+)
+
+type signatureMediaType string
+
+// signatureAlgorithm lists supported Signature algorithms.
+type signatureAlgorithm string
+
+const (
+	RSASSA_PSS_SHA_256 signatureAlgorithm = "RSASSA_PSS_SHA_256"
+	RSASSA_PSS_SHA_384 signatureAlgorithm = "RSASSA_PSS_SHA_384"
+	RSASSA_PSS_SHA_512 signatureAlgorithm = "RSASSA_PSS_SHA_512"
+	ECDSA_SHA_256      signatureAlgorithm = "ECDSA_SHA_256"
+	ECDSA_SHA_384      signatureAlgorithm = "ECDSA_SHA_384"
+	ECDSA_SHA_512      signatureAlgorithm = "ECDSA_SHA_512"
+)
+
+// SignerInfo represents a parsed signature envelope that is agnostic to signature envelope format.
+type SignerInfo struct {
+	Payload            []byte
+	PayloadContentType string
+	SignedAttributes   SignedAttributes
+	UnsignedAttributes UnsignedAttributes
+	SignatureAlgorithm signatureAlgorithm
+	CertificateChain   []x509.Certificate
+	Signature          []byte
+	TimestampSignature []byte
+}
+
+// SignedAttributes represents signed metadata in the Signature envelope
+type SignedAttributes struct {
+	SigningTime        time.Time
+	Expiry             time.Time
+	ExtendedAttributes []Attribute
+}
+
+// UnsignedAttributes represents unsigned metadata in the Signature envelope
+type UnsignedAttributes struct {
+	SigningAgent string
+}
+
+// SignRequest is used to generate Signature.
+type SignRequest struct {
+	Payload             []byte
+	PayloadContentType  string
+	CertificateChain    []x509.Certificate
+	SignatureProvider   SignatureProvider
+	SigningTime         time.Time
+	Expiry              time.Time
+	ExtendedSignedAttrs []Attribute
+	SigningAgent        string
+}
+
+type Attribute struct {
+	Key      string
+	Critical bool
+	Value    interface{}
+}
+
+// SignatureProvider is used to sign bytes generated after creating Signature envelope.
+type SignatureProvider interface {
+	Sign([]byte) ([]byte, error)
+}
+
+type SignatureEnvelope struct {
+	rawSignatureEnvelope []byte
+	internalEnvelope     internalSignatureEnvelope
+}
+
+// Contains a set of common methods that every Signature envelope format must implement.
+type internalSignatureEnvelope interface {
```

Comment:
I think given small codebase it might not be worth moving it to internal pkg but if in future we add more sig format we can do this refactoring..

Reply:
We may want to do It sooner so that we don't accidentally export an internal struct. We just need to move the internal types to their respective files in /internal

Reply:
As per conversation over call, we will move to internal package after introducing new sig format(cose?) if feel that the package bloated or if we want to share multiple method across multiple packages.